### PR TITLE
MULE-11292: Functions, TypedValues and ParameterResolvers are not sup…

### PR DESCRIPTION
**Pull request by estebanwasinger**
…ported inside a Pojo

### CODE REVIEW

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java [hunk 1]
```diff
@@ -188,22 +197,21 @@ protected ResolverSet getParametersAsResolverSet(ParameterizedModel model, List<
...
-              (List) ((Collection) value).stream().map(this::toValueResolver).collect(new ImmutableListCollector()));
+      resolver = getCollectionResolver((Collection) value, modelProperties);
     } else if (value instanceof Map) {
-      Map<Object, Object> map = (Map<Object, Object>) value;
-      Map<ValueResolver<Object>, ValueResolver<Object>> normalizedMap = new LinkedHashMap<>(map.size());
-      map.forEach((key, entryValue) -> normalizedMap.put((ValueResolver<Object>) toValueResolver(key),
-                                                         (ValueResolver<Object>) toValueResolver(entryValue)));
-      resolver = MapValueResolver.of(map.getClass(), copyOf(normalizedMap.keySet()), copyOf(normalizedMap.values()));
+      resolver = getMapResolver((Map<Object, Object>) value, modelProperties);
+    } else if (modelProperties.stream().anyMatch(modelProperty -> modelProperty instanceof ParameterResolverTypeModelProperty)) {
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java]: extract this into a method like isParameterResolver()

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java [hunk 2]
```diff
@@ -188,22 +197,21 @@ protected ResolverSet getParametersAsResolverSet(ParameterizedModel model, List<
...
+      resolver = new StaticValueResolver<>(new StaticParameterResolver<>(value));
+    } else if (modelProperties.stream().anyMatch(modelProperty -> modelProperty instanceof TypedValueTypeModelProperty)) {
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java]: idem

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java [hunk 3]
```diff
@@ -188,22 +197,21 @@ protected ResolverSet getParametersAsResolverSet(ParameterizedModel model, List<
...
+      resolver = getCollectionResolver((Collection) value, modelProperties);
     } else if (value instanceof Map) {
-      Map<Object, Object> map = (Map<Object, Object>) value;
-      Map<ValueResolver<Object>, ValueResolver<Object>> normalizedMap = new LinkedHashMap<>(map.size());
-      map.forEach((key, entryValue) -> normalizedMap.put((ValueResolver<Object>) toValueResolver(key),
-                                                         (ValueResolver<Object>) toValueResolver(entryValue)));
-      resolver = MapValueResolver.of(map.getClass(), copyOf(normalizedMap.keySet()), copyOf(normalizedMap.values()));
+      resolver = getMapResolver((Map<Object, Object>) value, modelProperties);
+    } else if (modelProperties.stream().anyMatch(modelProperty -> modelProperty instanceof ParameterResolverTypeModelProperty)) {
+      resolver = new StaticValueResolver<>(new StaticParameterResolver<>(value));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java]: how are you so sure you don't need a `ExpressionBasedParameterResolver`?

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java [hunk 4]
```diff
@@ -241,7 +249,7 @@ public void resolveParameters(ObjectType objectType, DefaultObjectBuilder builde
       Field objectField = getField(objectClass, key);
 
       if (parameters.containsKey(key)) {
-        valueResolver = toValueResolver(parameters.get(key));
+        valueResolver = toValueResolver(parameters.get(key), emptySet());
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java]: overload the method and have one which delegates with the emptySet()

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java [hunk 5]
```diff
@@ -513,89 +515,106 @@ protected ClassLoader getContextClassLoader() {
...
-    } else if (isParameterResolver(modelProperties) && value != null) {
-      resolver = new ExpressionBasedParameterResolverValueResolver<>((String) value, expectedType, muleContext);
-    } else if (isTypedValue(modelProperties)) {
-      resolver = new ExpressionTypedValueValueResolver((String) value, expectedClass, muleContext);
+    boolean isExpression = isExpression(value, parser);
+
+    if (isExpression) {
+      if (isExpressionFunction(modelProperties)) {
+        resolver = new ExpressionFunctionValueResolver<>((String) value, expectedType, muleContext);
+      } else if (isParameterResolver(modelProperties, expectedType)) {
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java]: this doesn't make sense... you already know if this is an expression or not. You shouldn't test it for parameter resolver twice (here and in line 539 just for selecting different resolvers). Use the isExpression variable in one and only if.

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java [hunk 6]
```diff
@@ -513,89 +515,106 @@ protected ClassLoader getContextClassLoader() {
...
+        resolver = new ExpressionBasedParameterResolverValueResolver((String) value, expectedType, muleContext);
+      } else if (isTypedValue(modelProperties, expectedType)) {
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/ExtensionDefinitionParser.java]: idem

File: modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java (hunk 3 above)
Comment (estebanwasinger) [inline modules/extensions-spring-support/src/main/java/org/mule/runtime/module/extension/internal/config/dsl/AbstractExtensionObjectFactory.java]: in that place only static values are resolved

**Pull request merged**